File: Algorithms/C++/Medium/116_Populating_Next_Right_Pointers_in_Each_Node.cpp
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        if (root == NULL) {
            return;
        }

        queue<TreeLinkNode*> q;
        q.push(root);

        while (!q.empty()) {
            int level_size = q.size();

            for (int i = 0; i < level_size; i++) {
                TreeLinkNode* node = q.front();
                q.pop();

                if (node->left != NULL) {
                    q.push(node->left);
                }
                if (node->right != NULL) {
                    q.push(node->right);
                }

                if (i < level_size - 1) {
                    node->next = q.front();
                }
            }
        }
    }
};

File: Algorithms/JavaScript/Medium/116_Populating_Next_Right_Pointers_in_Each_Node.js
/**
 * Definition for binary tree with next pointer.
 * function TreeLinkNode(val) {
 *     this.val = val;
 *     this.left = this.right = this.next = null;
 * }
 */

/**
 * @param {TreeLinkNode} root
 * @return {void} Do not return anything, modify tree in-place instead.
 */
var connect = function(root) {
    if (root === null) {
        return
    }

    var queue = []
    queue.push(root)

    while (queue.length > 0) {
        var currentLevelSize = queue.length
        for (var i = 0; i < currentLevelSize; i++) {
            var node = queue.shift()
            if (i < currentLevelSize - 1) {
                node.next = queue[0]
            }

            if (node.left !== null) {
                queue.push(node.left)
            }
            if (node.right !== null) {
                queue.push(node.right)
            }
        }
    }
};
